---
title: "Class 9: Stuctural Bioinformatics 1"
author: "Tahmid Ahmed"
format: gfm
---

## The RCSB Protein Data Bank (PDB)

Protein structures by X-ray crystalography dominate this database. We are skipping Q1-3 because the site was too slow for us.

>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

The hydrogen atoms are too small to image therefore for each water molecule we are seeing the oxygen atom.

>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

Water molecule 308. 

>Q6 Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![HIV-Pr structure from 1hsg](1HSG.png)

>Q6.2 Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

The protein could undergo a conformational change in response to the environment such as pH or specific ions, to allow larger substrates to fit into the active site. 

## 3. Introduction to Bio3D in R

Bio3D is an R package for structural bioinformatics. To use it we need to callit up with the `library()` function (just like any package).
We also use `read.pdb()` to read a PDB file.

```{r}
library(bio3d)
pdb <- read.pdb("1hsg")
pdb
```

>Q7: How many amino acid residues are there in this pdb object?

198 residues.

>Q8: Name one of the two non-protein residues?

HOH

>Q9: How many protein chains are in this structure?

2 chains being A and B.

>Q10. Which of the packages above is found only on BioConductor and not CRAN?

msa

>Which of the above packages is not found on BioConductor or CRAN?

Grantlab/bio3d-view under bitbucket

>True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

True

# Comparative analysis of Adenylate kinase (ADK)

We will start our analysis with a single PDB id (code from the PDB database): 1AKE

First we get it's primary sequence:

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

>Q13. How many amino acids are in this sequence, i.e. how long is this sequence?

214 amino acids.

```{r}
# Blast or hmmer search 
b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
hits <- plot(b)
```

```{r}
# List out some 'top hits'
head(hits$pdb.id)
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE)#, exefile="msa")
```

```{r}
pdbs <- pdbaln(files, fit = TRUE)#, exefile="msa")
```
```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
#plot too large to input
#plot(pdbs, labels=ids)
```


```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

```{r}
# NMA of all structures
modes <- nma(pdbs)
```

```{r}
plot(modes, pdbs, col=grps.rd)
```

>Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

I notice that the green line has the highest peaks. The colored lines have substantially higher peaks than the black lines. This indicates that adk has two conformations, one being the colored lines and the other being the black line. I believe they differ most around the 150 residue region. 

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

{{< video PC_1.PDB_animate-trajectory.mp4 >}}